### PR TITLE
Add utils.index_has_overlap()

### PR DESCRIPTION
**Pull request by Bahaabrougui**
image

### CODE REVIEW

**Comment by hagenw** [comment]
Could you please split this pull request up into two as it is helpful and good practice to focus on one feature per pull request.

So it seems starting with "Add audformat.utils.iter_index_by_file()" would be a good start.

**Comment by Bahaabrougui** [comment]
> Could you please split this pull request up into two as it is helpful and good practice to focus on one feature per pull request.
> 
> So it seems starting with "Add audformat.utils.iter_index_by_file()" would be a good start.

In this case do I have to implement a separate test for `iter_index_by_file()` too ? As the `has_overlap()` test covered it so far

**Comment by hagenw** [comment]
yes, would be good to have separate tests.

File: audformat/utils/__init__.py [hunk 1]
```diff
@@ -13,5 +14,5 @@
     replace_file_extension,
     to_filewise_index,
     to_segmented_index,
-    union,
+    union
```
Comment (frankenjoe) [inline audformat/utils/__init__.py:17]:
```suggestion
    union,
```

File: audformat/core/utils.py [hunk 2]
```diff
@@ -326,6 +326,34 @@ def expand_file_path(
...
+        index: pd.Index,
+) -> bool:
+    r"""Check if one or more segments overlap.
+
+    Args:
+        index: index object conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Returns:
+        boolean
```
Comment (frankenjoe) [inline audformat/core/utils.py:339]:
```suggestion
        ``True`` if overlap is detected, otherwise ``False``
```

File: audformat/core/utils.py [hunk 3]
```diff
@@ -326,6 +326,34 @@ def expand_file_path(
     return index
 
 
+def has_overlap(
```
Comment (frankenjoe) [inline audformat/core/utils.py:329]: Maybe this is a bit too generic, maybe we should change to `index_has_overlap`? @hagenw what do you think?

File: tests/test_utils.py [hunk 4]
```diff
@@ -624,6 +624,28 @@ def test_expand_file_path(tmpdir, index, root, expected):
     pd.testing.assert_index_equal(expanded_index, expected)
 
 
+@pytest.mark.parametrize(
+    'index, expected',
+    [
+        (
+            audformat.segmented_index(['f1'] * 2, [0, 1], [2, 3]),
```
Comment (frankenjoe) [inline tests/test_utils.py:631]: Could you please add tests for an empty index and a segmented index with two files where the segments across files overlap but not within files (to make sure we only detect overlap within a file).

Comment (frankenjoe) [reply tests/test_utils.py:631]: And maybe also a test without overlap but where the segments touch each other (we expect `False' in that case).

File: audformat/utils/__init__.py (hunk 1 above)
Comment (Bahaabrougui) [reply audformat/utils/__init__.py:17]: Any particular reason why we add comma at the end ? Is it also some standardization ?

File: audformat/core/utils.py (hunk 3 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:329]: I agree, let's change it to that so it's more clear

File: audformat/utils/__init__.py (hunk 1 above)
Comment (frankenjoe) [reply audformat/utils/__init__.py]: Yes, if we have multiple arguments / elements etc. over multiple lines we always end with an `,` after the last entry.

File: audformat/core/utils.py (hunk 3 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:329]: Or, why don't we keep the name as it is but add support for pd.Series and pd.Dataframe, just like we did with `iter_by_file` ?

File: tests/test_utils.py (hunk 4 above)
Comment (Bahaabrougui) [reply tests/test_utils.py:631]: > Could you please add tests for an empty index and a segmented index with two files where the segments across files overlap but not within files (to make sure we only detect overlap within a file).

Added.

> And maybe also a test without overlap but where the segments touch each other (we expect False in that case).

There is already a test case for that

File: audformat/core/utils.py (hunk 3 above)
Comment (frankenjoe) [reply audformat/core/utils.py:329]: When I see `audformat.utils.has_overlap(df)` it is not intuitive for a user to know what will happen. As a user I would probably assume it checks if values in the columns overlap and not in the index. So I would suggest to rename and only support index.

File: tests/test_utils.py (hunk 4 above)
Comment (frankenjoe) [reply tests/test_utils.py:631]: We usually leave it to the reviewer to unresolve the issues.

File: audformat/core/utils.py (hunk 3 above)
Comment (frankenjoe) [reply audformat/core/utils.py:329]: If we force the user to do `audformat.utils.index_has_overlap(df.index)` the code will be longer, but it's also clear what the statement does.

Comment (Bahaabrougui) [reply audformat/core/utils.py:329]: I added support for `Series` and `Dataframe`, if you think I should revert it to just `pd.Index` let me know

File: tests/test_utils.py (hunk 4 above)
Comment (frankenjoe) [reply tests/test_utils.py]: > There is already a test case for that

Sorry, overlooked it.

> Added.

Just to be on the safe side, please add also one for an empty `filewise` index.

File: audformat/core/utils.py (hunk 3 above)
Comment (frankenjoe) [reply audformat/core/utils.py:329]: Ok, as long as we rename to `index_has_overlap()` it is probably ok (and convenient) to support `pd.Series` and `pd.DataFrame` as we do with other functions. Please extend the tests accordingly.

File: audformat/core/utils.py [hunk 5]
```diff
@@ -326,6 +326,36 @@ def expand_file_path(
...
+        obj: typing.Union[pd.Index, pd.DataFrame, pd.Series],
+) -> bool:
+    r"""Check if one or more segments overlap.
```
Comment (frankenjoe) [inline audformat/core/utils.py:332]:
```suggestion
    r"""Check if one or more segments in the index overlap.
```

Since we now support `Series` and `DataFrame` I think it's better to be more precise here.

File: docs/api-utils.rst [hunk 6]
```diff
@@ -19,6 +19,11 @@ expand_file_path
 
 .. autofunction:: expand_file_path
 
+has_overlap
+___________
```
[the remark this answers is not in the document]
Comment (frankenjoe) [reply docs/api-utils.rst:23]: Otherwise it will be treated as another sub level I think.

File: tests/test_utils.py [hunk 7]
```diff
@@ -624,6 +624,52 @@ def test_expand_file_path(tmpdir, index, root, expected):
...
+            ),
+            False,
+        ),
+        (
+            pd.Series(
+                index=audformat.segmented_index(
+                    ['f1'] * 2,
+                    [0, 2],
+                    [2, 3],
+                )
```
Comment (frankenjoe) [inline tests/test_utils.py:656]:
```suggestion
                ),
```

File: tests/test_utils.py [hunk 8]
```diff
@@ -624,6 +624,52 @@ def test_expand_file_path(tmpdir, index, root, expected):
...
+            ),
+            False,
+        ),
+        (
+            pd.DataFrame(
+                index=audformat.filewise_index(['f1'] * 2)
```
Comment (frankenjoe) [inline tests/test_utils.py:662]: I would prefer to keep the`filewise` index test as before and add one for `DataFrame` in addition.

File: tests/test_utils.py (hunk 4 above)
Comment (Bahaabrougui) [reply tests/test_utils.py]: > We usually leave it to the reviewer to unresolve the issues.

Okay didn't know that

> Just to be on the safe side, please add also one for an empty filewise index.
I will

File: audformat/core/utils.py (hunk 3 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:329]: Oh the page didn't reload and didn't see your comments on not adding support for Dataframes and Series. But with the last comment I guess should be fine, I'll just rename it so it's more clear

File: tests/test_utils.py (hunk 4 above)
Comment (Bahaabrougui) [reply tests/test_utils.py]: >  We usually leave it to the reviewer to unresolve the issues.

Just do double check, you mean resolve right ?

**Comment by Bahaabrougui** [comment]
@frankenjoe, all suggestions are applied.

File: tests/test_utils.py (hunk 4 above)
Comment (frankenjoe) [reply tests/test_utils.py]: Oh of course I mean "resolve" :) This way you give the reviewer the chance to check your proposed changes and if necessary complain again :)

Comment (frankenjoe) [reply tests/test_utils.py]: We usually still leave a comment for the reviewer like "suggestion applied" or something, so the reviewer knows that the issue has been tackled.

Comment (Bahaabrougui) [reply tests/test_utils.py]: > Oh of course I meant "resolve" :) This way you give the reviewer the chance to check your proposed changes and if necessary complain again :)

Haha okay, I get it :)

Comment (Bahaabrougui) [reply tests/test_utils.py]: suggestion applied :)

File: audformat/core/utils.py [hunk 9]
```diff
@@ -353,6 +353,36 @@ def hash(
...
+        ``True`` if overlap is detected, otherwise ``False``
+
+    """
+
+    index = obj if isinstance(obj, pd.Index) else obj.index
+
+    if index_type(index) == define.IndexType.FILEWISE:
+        return False
+
+    index = to_segmented_index(index, allow_nat=False)
```
Comment (frankenjoe) [inline audformat/core/utils.py:375]: Mhh, I'm not sure if we should force `allow_nat=False` here as it can be costly. Instead we should probably handle `NaT`. E.g.:

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [1, NaT],
)
```

-> False

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [NaT, 3],
)
```

-> True

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [NaT, NaT],
)
```

-> True

Comment (Bahaabrougui) [reply audformat/core/utils.py:375]: Hm why and how costly it is ?

Comment (frankenjoe) [reply audformat/core/utils.py:375]: It will replace `NaT` with the duration of the file, so it has to get the duration for every file that is affected.

Comment (frankenjoe) [reply audformat/core/utils.py:375]: In fact, it's not necessary to replace `NaT` with an exact value anyway. The result will be the same as long as we replace it with a large enough value. So I recommend to remove the line and instead insert:

```
ends = ends.fillna(pd.to_timedelta(sys.maxsize))
```

before the `if` condition within the loop.

Comment (Bahaabrougui) [reply audformat/core/utils.py]: Yes I agree this is smarter way to do it. I applied the suggestions and added the adequate tests.

File: audformat/core/utils.py [hunk 10]
```diff
@@ -372,11 +373,12 @@ def index_has_overlap(
     if index_type(index) == define.IndexType.FILEWISE:
         return False
 
-    index = to_segmented_index(index, allow_nat=False)
+    index = to_segmented_index(index)
```
Comment (frankenjoe) [inline audformat/core/utils.py:376]: Can be removed as we already know that it must be a segmented index.

File: audformat/core/utils.py (hunk 9 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: By the way from where can I get the knowledge of what's faster and what's not and these kind of optimizations when handling data structures ? I mean I find things on internet from time to time but maybe you have a better source

File: audformat/core/utils.py (hunk 10 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:376]: Yea already did, see latest commit

File: tests/test_utils.py [hunk 11]
```diff
@@ -7,6 +7,8 @@
 import pytest
 
 import audeer
+from pandas._libs.tslibs.nattype import NaT
```
Comment (frankenjoe) [inline tests/test_utils.py:10]: I recommend to remove this important and use `pd.NaT` in the tests below.

File: audformat/core/utils.py (hunk 10 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:376]: What if the user passes a data frame that index ain't conform to audformat ? This keeps us in the safe side, or is it also a costly operation ?

Comment (frankenjoe) [reply audformat/core/utils.py:376]: Not really, since `to_segmented_index()` also expects an index that is conform to audformat.

File: tests/test_utils.py (hunk 11 above)
Comment (Bahaabrougui) [reply tests/test_utils.py:10]: Okay, why the recommendation tho ?

File: audformat/core/utils.py (hunk 9 above)
Comment (frankenjoe) [reply audformat/core/utils.py]: Any operation that needs to access information from disk becomes costly once you scale it up. Otherwise it's not always obvious what will be costly and what not. So usually the only way to find a bottleneck is by running your function on a large number of files / segments. With `audb` we also used a profiler at some point to nail down the expensive operations.

File: audformat/core/utils.py (hunk 10 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: Ok, applied.

File: tests/test_utils.py (hunk 11 above)
Comment (frankenjoe) [reply tests/test_utils.py]: 1. We already import `pandas`, so there is no need for an extra import
2. You are importing from a private module (`_libs`) here, which is not a good idea. They may change it in a future version, but `pd.NaT` will always be there.

File: audformat/core/utils.py [hunk 12]
```diff
@@ -353,6 +354,36 @@ def hash(
...
+    r"""Check if one or more segments in the index overlap.
+
+    Args:
+        obj: object conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Returns:
+        ``True`` if overlap is detected, otherwise ``False``
+
+    """
```
Comment (frankenjoe) [inline audformat/core/utils.py:369]: I think we should add a docstring example.

File: audformat/core/utils.py (hunk 9 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: Hm okay, a profiler is a tool that helps you simulate bottlenecks and gives you statistics and stuff right ? And what did you once you figured what are the expensive operations ? Some optimizations or just noting down that they should be avoided on large structures ?

File: tests/test_utils.py (hunk 11 above)
Comment (Bahaabrougui) [reply tests/test_utils.py]: Okay understandable thanks, applied.

File: audformat/core/utils.py (hunk 9 above)
Comment (frankenjoe) [reply audformat/core/utils.py]: > Hm okay, a profiler is a tool that helps you simulate bottlenecks and gives you statistics and stuff right ?

Yes, a profiler shows you how long each function call takes and helps you to identify the critical calls

> And what did you once you figured what are the expensive operations?

Try to optimize you code by avoiding expensive operations :) E.g. here we now replace `NaT` with a large value which is much faster than accessing the actual file duration. But the result will be the same. These kind of things.

File: audformat/core/utils.py (hunk 12 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py:381]: Added, and description above updated.

File: audformat/core/utils.py (hunk 9 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: Okay perfect, thanks for explaining. Feel free to resolve the thread.

File: audformat/core/utils.py (hunk 12 above)
Comment (frankenjoe) [reply audformat/core/utils.py:381]: Cool, thanks. I have a slight suggestion though:

```
    Example:
        >>> index = filewise_index(['f1', 'f2'])
        >>> index_has_overlap(index)
        False
        >>> index = segmented_index(
        ...        ['f1', 'f2'],
        ...        [0, 1],
        ...        [2, 3],
        ...    )
        >>> index_has_overlap(index)
        False
        >>> index = segmented_index(
        ...        ['f1'] * 2,
        ...        [0, 1],
        ...        [2, 3],
        ...    )
        >>> index_has_overlap(index)
        True
```

Why? We now start with the simplest example and move on to advanced ones.

File: audformat/core/utils.py [hunk 13]
```diff
@@ -353,6 +354,48 @@ def hash(
...
+        ...        [2, 3],
+        ...    )
+        >>> index_has_overlap(index)
+        True
+        >>> index = filewise_index(['f1', 'f2'])
+        >>> index_has_overlap(index)
+        False
+
+    """
+
```
Comment (frankenjoe) [inline audformat/core/utils.py:382]: We usually don't start with a blank line after the docstring.

File: audformat/core/utils.py (hunk 12 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: Make sense, applied.

Comment (frankenjoe) [reply audformat/core/utils.py]: Thanks, could you please update the description then we are good to go I think.

Comment (frankenjoe) [reply audformat/core/utils.py]: Please also update the title to `Add utils.index.has_overlap()` (no need to mention that we add tests).

Comment (Bahaabrougui) [reply audformat/core/utils.py]: > Thanks, could you please update the description then we are good to go I think.

Already did

> Please also update the title to Add utils.index_has_overlap() (no need to mention that we add tests).

Done now

Comment (frankenjoe) [reply audformat/core/utils.py]: > Already did

It still shows the old example, but ok, not so important.

**Comment by frankenjoe** [comment]
@hagenw no more comments from my side

File: audformat/core/utils.py (hunk 12 above)
Comment (Bahaabrougui) [reply audformat/core/utils.py]: Weird, are you talking about the PR description? Because it's updated for me 

image

Comment (frankenjoe) [reply audformat/core/utils.py]: Ok, now for me, too. Maybe it was just a caching issue. Looks great now!

**Comment by hagenw** [comment]
If you have an index with the same files, results are not as expected:

```python
>>> index = audformat.filewise_index(['f1', 'f1'])
>>> audformat.utils.index_has_overlap(index)
False
```

**Comment by frankenjoe** [comment]
> If you have an index with the same files, results are not as expected:

Well, one could argue such an index is not conform to audformat as we don't allow duplicates.

**Comment by hagenw** [comment]
> > If you have an index with the same files, results are not as expected:
> 
> Well, one could argue such an index is not conform to `audformat` as we don't allow duplicates.

That's true, but as we don't raise an error it's still risky that it returns the wrong value.

**Comment by frankenjoe** [comment]
> That's true, but as we don't raise an error it's still risky that it returns the wrong value.

We have [assert_no_duplicates()](https://audeering.github.io/audformat/api.html#assert-no-duplicates), but at some point decided it to leave it to the user to call it since it was too expensive to call it in every function. So I don't think we should start calling it again in this function.

**Comment by hagenw** [comment]
> > That's true, but as we don't raise an error it's still risky that it returns the wrong value.
> 
> We have [assert_no_duplicates()](https://audeering.github.io/audformat/api.html#assert-no-duplicates), but at some point decided it to leave it to the user to call it since it was too expensive to use it in every function. So I don't think we should start calling it again in this function.

In this case we only need to check for duplicates in filewise index and can return a `True` as soon as we find one duplicated entry, do you think there is no fast way to check this?
Another solution would be to say that for a filewise index it always returns `False` as there are no segments in a filewise index.

You also state in the docstring at the moment:
"Check if one or more segments in the index overlap."

So maybe just add another sentence, that for a filewise index it always returns `False`.

**Comment by frankenjoe** [comment]
> So maybe just add another sentence, that for a filewise index it always returns `False`.

Yes, let's do that.

**Comment by Bahaabrougui** [comment]
> > > That's true, but as we don't raise an error it's still risky that it returns the wrong value.
> > 
> > 
> > We have [assert_no_duplicates()](https://audeering.github.io/audformat/api.html#assert-no-duplicates), but at some point decided it to leave it to the user to call it since it was too expensive to use it in every function. So I don't think we should start calling it again in this function.
> 
> In this case we only need to check for duplicates in filewise index and can return a `True` as soon as we find one duplicated entry, do you think there is no fast way to check this? Another solution would be to say that for a filewise index it always returns `False` as there are no segments in a filewise index.
> 
> You also state in the docstring at the moment: "Check if one or more segments in the index overlap."
> 
> So maybe just add another sentence, that for a filewise index it always returns `False`.

I've updated the description.

File: audformat/core/utils.py [hunk 14]
```diff
@@ -353,6 +354,56 @@ def hash(
     return str(pd.util.hash_pandas_object(obj).sum())
 
 
+def index_has_overlap(
+        obj: typing.Union[pd.Index, pd.DataFrame, pd.Series],
+) -> bool:
+    r"""Check if one or more segments in the index overlap.
+
+    If the index is filewise, the return value will always be False.
```
[the remark this answers is not in the document]
Comment (frankenjoe) [reply audformat/core/utils.py]: > Does it mean I have to unresolve it after it happens automatically ?

If you apply a suggestion by the reviewer directly, there is no need to keep it open.

**Comment by frankenjoe** [comment]
Since it references an internal project, I suggest we remove this from the description:

image

**Comment by Bahaabrougui** [comment]
> Since it references an internal project, I suggest we remove this from the description:
> 
> image

Ah true forget this was supposed to be opensource :D

**Comment by hagenw** [comment]
Great, and thanks for updating this.

BTW, the corresponding segmented index works, which is nice to have:

```python
>>> index = audformat.filewise_index(['f1', 'f1'])
>>> index = audformat.utils.to_segmented_index(index)
>>> index
MultiIndex([('f1', '0 days', NaT),
            ('f1', '0 days', NaT)],
           names=['file', 'start', 'end'])
>>> audformat.utils.index_has_overlap(index)
True
>>> index = audformat.filewise_index(['f1', 'f2'])
>>> index = audformat.utils.to_segmented_index(index)
>>> index
MultiIndex([('f1', '0 days', NaT),
            ('f2', '0 days', NaT)],
           names=['file', 'start', 'end'])
>>> audformat.utils.index_has_overlap(index)
False
```
